docs/docs/reference/rendering-options/server-side-rendering.md

```diff
@@ -0,0 +1,133 @@
+as server-rendered.
+
+The `context` parameter is an object with the following keys:
+
+- `headers`: The request headers
+- `method`: The request method, e.g. `GET`
+- `url`: The request URL
+- `query`: An object representing the query string
+- `params`: If you use [File System Route API](/docs/reference/routing/file-system-route-api/) the URL path gets passed in as `params`. For example, if your page is at `src/pages/{Product.name}.js` the `params` will be an object like `{ name: 'value' }`.
+
+`getServerData` can return an object with two keys:
+
+- `props` (optional): Object containing the data passed to `serverData` page prop. Should be a serializable object.
+- `headers` (optional): Object containing `headers` that are sent to the browser and caching proxies/CDNs, e.g. cache headers
+
+### Use `serverData` prop in React page component
+
+The `props` object you return from `getServerData` gets passed to the page component as `serverData` prop.
+
+```js
+import * as React from "react"
+
+const Page = ({ serverData }) => {
+  const { dogImage } = serverData
+
+  // Use dogImage in your page...
+}
+
+export async function getServerData() {
+  const res = await fetch(`https://a-dog-image.com/random`)
+  const data = await res.json()
+
+  return {
+    props: {
+      dogImage: data,
+    },
+  }
+}
+
+export default Page
+```
+
+## Interplay with build-time GraphQL queries
+
+Server-rendered pages also support regular Gatsby GraphQL page queries. The page query is executed at build-time
+and the data is passed to React component as `data` prop on each render (along with `serverData` prop).
+
+Keep in mind that `data` will be the same every time but `serverData` will change according to return value of your `getServerData` function.
+Runtime GraphQL queries are not supported yet.
+
+```js
+import * as React from "react"
+
+const Page = ({ data, serverData }) => {
+  const { site } = data
+  const { dogImage } = serverData
+  // Use dogImage and site info in your page...
+}
+
+export const pageQuery = graphql`
+  query PageData {
+    site {
+      siteMetadata {
+        title
+      }
+    }
+  }
+`
+
+export async function getServerData() {
+  const res = await fetch(`https://a-dog-image.com/random`)
+  const data = await res.json()
+
+  return {
+    props: {
+      dogImage: data,
+    },
+  }
+}
+
+export default Page
+```
+
+## Working with server-rendered pages locally
+
+Server-rendered pages work with both `gatsby develop` and `gatsby serve`. Page will be
+re-generated on each request.
+
+## Using in production
+
+Server-side rendering requires a running NodeJS server. You can put NodeJS running `gatsby serve`
+behind a CDN, however it requires additional infrastructure: monitoring, logging, crash-recovery, etc.
+
+Complete setup with auto-scaling is available for you in [Gatsby Cloud](/products/cloud/) out-of-the-box.
+
+## How it works
```

Comment:
Similar note here. This section would benefit from Flo's diagram and a list of ordered steps. We can add that in a follow-up PR later.